src/main/scala/com/nec/ve/GroupingFunction.scala

```diff
@@ -100,118 +100,57 @@ object GroupingFunction {
     }
   }
 
-  def groupData(data: List[DataDescription], totalBuckets: Int) = {
-    if (data.count(_.keyOrValue.isKey) <= 0) {
-      /*
-        In the case where there are no input key columns, we simply clone the
-        input data to the output pointers.
-       */
-      groupDataNoKeyColumns(data)
-
-    } else {
-      groupDataNormal(data, totalBuckets)
-    }
-  }
-
-  def groupDataNoKeyColumns(data: List[DataDescription]): CFunction2 = {
-    val inputs = data.zipWithIndex.map { case (DataDescription(veType, isKey), idx) =>
-      veType.makeCVector(s"${isKey.renderValue}_${idx}")
-    }
-
-    val outputs = data.zipWithIndex.map { case (DataDescription(veType, isKey), idx) =>
-      veType.makeCVector(s"output_${isKey.renderValue}_${idx}")
-    }
-
-    val arguments = inputs.map(PointerPointer(_)) ++
-      List(CFunctionArgument.Raw("int* sets")) ++
-      outputs.map(PointerPointer(_))
-
-    CFunction2(
-      arguments,
-      CodeLines.from(
-        "// Write out the number of buckets",
-        s"sets[0] = 1;",
-        "",
-        (outputs, inputs).zipped.map(cloneCVecStmt(_, _))
-      )
-    )
-  }
-
-  def copyVecToBucketsStmt(
-    output: CVector,
-    input: CVector,
-    buckets: Int,
-    idToBucket: String,
-    bucketToCount: String
-  ): CodeLines = {
-    require(
-      output.veType == input.veType,
-      "Cannot copy cVector values to their buckets - input and output VeTypes are not the same!"
-    )
-
+  private[ve] def copyVecToBucketsStmt(output: CVector, input: CVector): CodeLines = {
     CodeLines.scoped(
       s"Copy elements of ${input.name}[0] to their respective buckets in ${output.name}"
     ) {
       CodeLines.from(
         // Perform the bucketing and get back T** (array of pointers)
-        s"auto ** tmp = ${input.name}[0]->bucket(${bucketToCount}, ${idToBucket});",
+        s"auto ** tmp = ${input.name}[0]->bucket(${GroupingFunction.GroupCountsId}, ${GroupingFunction.GroupAssignmentsId});",
         /*
           Allocate the nullable_T_vector[] with size buckets
 
-          NOTE: This cast should not be correct, because we are allocating a T*
-          array (T**) but type-casting it to T*.  However, for some reason,
-          fixing this will lead an invalid free() later on.  Will need to
-          investigate fix this.
+          NOTE: This cast is incorrect, because we are allocating a T* array
+          (T**) but type-casting it to T*.  However, for some reason, fixing
+          this will lead an invalid free() later on.  Will need to investigate
+          and fix this in the future.
```

Comment:
Add: "This is likely due to an error in the JavaCpp function definition."

Reply:
Done